std/algorithm/comparison.d

```diff
@@ -732,75 +731,130 @@ template equal(alias pred = "a == b")
     See_Also:
         $(HTTP sgi.com/tech/stl/_equal.html, STL's _equal)
     +/
-    bool equal(Range1, Range2)(Range1 r1, Range2 r2)
-    if (isInputRange!Range1 && isInputRange!Range2 &&
-        is(typeof(binaryFun!pred(r1.front, r2.front))))
+    bool equal(R, Ss...)(R r, Ss ss)
+    if (ss.length >= 1 &&
+        allSatisfy!(isInputRange, R, Ss) &&
+        (!isInfinite!R || !allSatisfy!(isInfinite, Ss)) // at least one must be finite
+        )
     {
-        static assert(!(isInfinite!Range1 && isInfinite!Range2),
-            "Both ranges are known to be infinite");
-
-        //No pred calls necessary
-        static if (isEmptyRange!Range1 || isEmptyRange!Range2)
-        {
-            return r1.empty && r2.empty;
-        }
-        else static if ((isInfinite!Range1 && hasFixedLength!Range2) ||
-            (hasFixedLength!Range1 && isInfinite!Range2))
-        {
-            return false;
-        }
-        //Detect default pred and compatible dynamic array
-        else static if (is(typeof(pred) == string) && pred == "a == b" &&
-            isArray!Range1 && isArray!Range2 && is(typeof(r1 == r2)))
-        {
-            return r1 == r2;
-        }
-        // if one of the arguments is a string and the other isn't, then auto-decoding
-        // can be avoided if they have the same ElementEncodingType
-        else static if (is(typeof(pred) == string) && pred == "a == b" &&
-            isAutodecodableString!Range1 != isAutodecodableString!Range2 &&
-            is(ElementEncodingType!Range1 == ElementEncodingType!Range2))
+        enum isEquableToR(T) = is(typeof({ return R.init == T.init; }));
+
+        // start by detecting default pred and compatible dynamic array
+        static if (is(typeof(pred) == string) &&
+                   pred == "a == b" &&
+                   isArray!R &&
+                   allSatisfy!(isArray, Ss) &&
+                   allSatisfy!(isEquableToR, Ss))
         {
-            import std.utf : byCodeUnit;
-
-            static if (isAutodecodableString!Range1)
-            {
-                return equal(r1.byCodeUnit, r2);
-            }
-            else
+            foreach (ref s; ss)
             {
-                return equal(r2.byCodeUnit, r1);
+                if (r != s) return false;
             }
+            return true;
         }
-        //Try a fast implementation when the ranges have comparable lengths
-        else static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
+        // use fast implementation when the ranges have comparable lengths
+        else static if (allSatisfy!(hasLength, R, Ss))
```

Comment:
I think here the logic would be - if any subset of size greater than 1 has length, compare the lengths within that subset.

Reply:
Do you mean something like

`
            for (; !r.empty; r.popFront())  // for each element in first range `r`
            {
                static foreach (s; rs) // for each other range `s`
                {
                    if (!binaryFun!(pred)(r.front, s.front)) return false;
                    + static if (!hasLength!(typeof(s)))
                    + {
                    +     if (s.empty) return false;
                    + }

                    s.popFront();
                }
            }

`

inside the "use fast implementation"?